File: internal/objectstore/objectstore.go
package objectstore

import (
	"context"
	"errors"
)

var ErrNotFound = errors.New("object not found")

type PutOptions struct {
	ContentType string
	Metadata    map[string]string
}

type ListOptions struct {
	MaxKeys           int
	ContinuationToken string
}

type ObjectInfo struct {
	Key      string
	Size     int64
	Metadata map[string]string
}

type ListPage struct {
	Keys                  []string
	IsTruncated           bool
	NextContinuationToken string
}

type ObjectStore interface {
	PutObject(ctx context.Context, key string, data []byte, opts PutOptions) error
	GetObject(ctx context.Context, key string) ([]byte, error)
	HeadObject(ctx context.Context, key string) (ObjectInfo, error)
	ListPrefix(ctx context.Context, prefix string, opts ListOptions) ([]string, error)
	ListPrefixPage(ctx context.Context, prefix string, opts ListOptions) (ListPage, error)
	DeleteObject(ctx context.Context, key string) error
}

func IsNotFound(err error) bool {
	return errors.Is(err, ErrNotFound)
}

func DeletePrefix(ctx context.Context, store ObjectStore, prefix string) (int, error) {
	deleted := 0
	for {
		page, err := store.ListPrefixPage(ctx, prefix, ListOptions{MaxKeys: 1000})
		if err != nil {
			return deleted, err
		}
		if len(page.Keys) == 0 {
			return deleted, nil
		}
		for _, key := range page.Keys {
			if err := store.DeleteObject(ctx, key); err != nil && !IsNotFound(err) {
				return deleted, err
			}
			deleted++
		}
	}
}

File: android-client/app/src/main/java/io/s3s5/android/objectstore/ObjectStore.kt
package io.s3s5.android.objectstore

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.TreeMap
import java.util.concurrent.atomic.AtomicLong

class ObjectNotFoundException(key: String) : Exception("object not found: $key")

data class PutOptions(
    val contentType: String = "",
    val metadata: Map<String, String> = emptyMap(),
)

data class ListOptions(
    val maxKeys: Int = 0,
)

data class ObjectInfo(
    val key: String,
    val size: Long,
    val metadata: Map<String, String> = emptyMap(),
)

interface ObjectStore {
    suspend fun putObject(key: String, data: ByteArray, options: PutOptions = PutOptions())
    suspend fun getObject(key: String): ByteArray
    suspend fun headObject(key: String): ObjectInfo
    suspend fun listPrefix(prefix: String, options: ListOptions = ListOptions()): List<String>
    suspend fun deleteObject(key: String)
}

data class StoreCounters(
    val put: Long,
    val get: Long,
    val head: Long,
    val list: Long,
    val delete: Long,
)

class CountingObjectStore(private val delegate: ObjectStore) : ObjectStore {
    private val putCount = AtomicLong()
    private val getCount = AtomicLong()
    private val headCount = AtomicLong()
    private val listCount = AtomicLong()
    private val deleteCount = AtomicLong()

    fun snapshot(): StoreCounters = StoreCounters(
        put = putCount.get(),
        get = getCount.get(),
        head = headCount.get(),
        list = listCount.get(),
        delete = deleteCount.get(),
    )

    override suspend fun putObject(key: String, data: ByteArray, options: PutOptions) {
        putCount.incrementAndGet()
        delegate.putObject(key, data, options)
    }

    override suspend fun getObject(key: String): ByteArray {
        getCount.incrementAndGet()
        return delegate.getObject(key)
    }

    override suspend fun headObject(key: String): ObjectInfo {
        headCount.incrementAndGet()
        return delegate.headObject(key)
    }

    override suspend fun listPrefix(prefix: String, options: ListOptions): List<String> {
        listCount.incrementAndGet()
        return delegate.listPrefix(prefix, options)
    }

    override suspend fun deleteObject(key: String) {
        deleteCount.incrementAndGet()
        delegate.deleteObject(key)
    }
}

class MemoryObjectStore : ObjectStore {
    private data class Entry(val data: ByteArray, val metadata: Map<String, String>)

    private val mutex = Mutex()
    private val objects = TreeMap<String, Entry>()

    override suspend fun putObject(key: String, data: ByteArray, options: PutOptions) {
        mutex.withLock {
            objects[key] = Entry(data.copyOf(), options.metadata.toMap())
        }
    }

    override suspend fun getObject(key: String): ByteArray = mutex.withLock {
        objects[key]?.data?.copyOf() ?: throw ObjectNotFoundException(key)
    }

    override suspend fun headObject(key: String): ObjectInfo = mutex.withLock {
        val entry = objects[key] ?: throw ObjectNotFoundException(key)
        ObjectInfo(key = key, size = entry.data.size.toLong(), metadata = entry.metadata)
    }

    override suspend fun listPrefix(prefix: String, options: ListOptions): List<String> = mutex.withLock {
        val keys = objects.keys
            .asSequence()
            .filter { it.startsWith(prefix) }
            .sorted()
        if (options.maxKeys > 0) {
            keys.take(options.maxKeys).toList()
        } else {
            keys.toList()
        }
    }

    override suspend fun deleteObject(key: String) {
        mutex.withLock {
            objects.remove(key)
        }
    }
}
